x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -139,7 +158,49 @@ export class AuthenticationService {
       return t.notHandled();
     });
 
-    this.logger.debug('Successfully registered core authentication handler.');
+    http.registerOnPreResponse(async (request, preResponse, toolkit) => {
+      if (preResponse.statusCode !== 401 || !canRedirectRequest(request)) {
+        return toolkit.next();
+      }
+
+      if (!this.authenticator) {
+        // Core doesn't allow returning error here.
+        this.logger.error('Authentication sub-system is not fully initialized yet.');
+        return toolkit.next();
+      }
+
+      // If users can eventually re-login we want to redirect them directly to the page they tried
+      // to access initially, but we only want to do that for routes that aren't part of the various
+      // authentication flows that wouldn't make any sense after successful authentication.
+      const originalURL = !request.route.options.tags.includes(ROUTE_TAG_AUTH_FLOW)
+        ? this.authenticator.getRequestOriginalURL(request)
+        : `${http.basePath.get(request)}/`;
+      if (!isLoginPageAvailable) {
+        return toolkit.render({
+          body: renderUnauthenticatedPage({ buildNumber, basePath: http.basePath, originalURL }),
+          headers: { 'Content-Security-Policy': http.csp.header },
+        });
+      }
+
+      const needsToLogout = (await this.session?.getSID(request)) !== undefined;
+      if (needsToLogout) {
```

Comment:
__note:__ unfortunately if we need to do a SLO then neither of our `unauthenticated` screens/messages will be shown in the result.